Refactor couch_httpd for before_response hook

jaydoane opened the pull request:
The main idea is to reduce duplicate code, and to create a single `do/5` function where the `before_response` hook is invoked, in addition to common logging, etc. I didn't know how to create funs for mochiweb parameterized module "methods", so I wrote simple wrappers ending in "_" to avoid collisions with existing functions.

@iilyak, can you LMK what you think of this effort so far? If you think it is an improvement, I could do a similar refactor of chttpd.erl

iilyak commented on src/couch_httpd.erl:
```diff
@@ -648,32 +648,32 @@ log_request(#httpd{mochi_req=MochiReq,peer=Peer}=Req, Code) ->
+start_response_length(Req, Code, Headers0, Length) ->
+    couch_log:notice("**** start_response_length", []),
+    Headers = start_response_headers(Req, Headers0, Length),
+    {Resp,_,_,_} = do(fun start_response_/2, Req, Code, Headers, undefined),
```
do is not descriptive enough to convey the purpose.

iilyak replied:
Would you mind adding spaces after comma in left hand side of match expression?
I.e. `{Resp, _, _, _}`.

iilyak commented on src/couch_httpd.erl:
```diff
@@ -761,13 +776,14 @@ send_json(Req, Code, Headers, Value) ->
-
-    Headers2 = Headers1 ++ server_header() ++
-               couch_httpd_auth:cookie_auth_header(Req, Headers1),
+add_headers(Req, Headers0) ->
```
You are changing the order of headers here. This is alright. But please make sure that all tests pass. I remember seeing few tests which do not use `?assertEqual(lists:usort(Expected), lists:usort(Headers))`. But rely on literal equivalence (`?assertEqual(Expected, Headers)`).

iilyak commented on src/couch_httpd.erl:
```diff
@@ -648,32 +648,32 @@ log_request(#httpd{mochi_req=MochiReq,peer=Peer}=Req, Code) ->
     end,
     {ok, Resp}.
 
+start_response_headers(Req, Headers0, Length) ->
```
Would it be more readable:

```
start_response_headers(Req, Headers0, undefined) ->
    basic_headers(Req, Headers0);
start_response_headers(Req, Headers0, Length) ->
   Headers = basic_headers(Req, Headers0),
   include_content_length_header(Headers, Length).
```

Feel free to ignore.

iilyak commented on src/couch_httpd.erl:
```diff
@@ -648,32 +648,32 @@ log_request(#httpd{mochi_req=MochiReq,peer=Peer}=Req, Code) ->
+    start_response_length(Req, Code, Headers, undefined).
+
+start_response_length(Req, Code, Headers0, Length) ->
+    couch_log:notice("**** start_response_length", []),
```
Don't forget to remove log statements before merge.

iilyak commented on src/couch_httpd.erl:
```diff
@@ -720,26 +738,23 @@ last_chunk(Resp) ->
+    Body0 = make_body({Type, Value}),
+    Headers1 = add_headers(Req, Headers0),
+    {Resp, Code, _, Body} = do(fun respond_/2, Req, Code0, Headers1, Body0),
+    maybe_log(Code, Body),
```
How about naming it `maybe_log_response`?

jaydoane commented:
@iilyak, I'm currently attempting to fix failing couchdb_cors_tests and couchdb_http_proxy_tests. There seem to be some subtle issues with this refactor which are causing a substantial number of failures, as you predicted :(

jaydoane commented:
@ilyak, I made all the changes you recommended that were still relevant after bug fixes. All tests now pass. Shall I squash into one commit?

iilyak commented:
@jaydoane: Please do. We might use this PR as a backup plan.

jaydoane committed:
Refactor couch_httpd

Create single point of entry for before_response plugin,
while reducing duplicate code.

jaydoane commented:
@iilyak: squashed.

jaydoane commented:
Closing in favor https://github.com/apache/couchdb-couch/pull/150
